File: backend/app/services/classification/document_type_gl_defaults.py
"""Finance-aligned default GL account names per playbook profile."""

from __future__ import annotations

from app.schemas.rule_book_config import ChartOfAccountEntry

# Playbook profile → suggested main ledger account name (must exist in tenant COA).
SUGGESTED_LEDGER_BY_PLAYBOOK_PROFILE: dict[str, str] = {
    "po_goods": "Raw Materials",
    "import_dossier": "Raw Materials",
    "freight_logistics": "Logistics",
    "standard_transactional": "Operating Expenses",
    "credit_adjustment": "Operating Expenses",
    "debit_note": "Operating Expenses",
    "direct_expense": "Operating Expenses",
    "employee_claim": "Travel Expense",
    "ar_goods": "Operating Revenue",
    "pre_transactional": "Suspense Account",
}

COA_TYPES_BY_PLAYBOOK_PROFILE: dict[str, tuple[str, ...]] = {
    "po_goods": ("Expense", "Asset"),
    "import_dossier": ("Expense", "Asset"),
    "freight_logistics": ("Expense",),
    "standard_transactional": ("Expense",),
    "credit_adjustment": ("Expense",),
    "debit_note": ("Expense",),
    "direct_expense": ("Expense",),
    "employee_claim": ("Expense",),
    "ar_goods": ("Revenue",),
    "pre_transactional": ("Liability",),
}


def suggested_ledger_for_playbook_profile(playbook_profile: str | None) -> str:
    token = (playbook_profile or "").strip().lower()
    return SUGGESTED_LEDGER_BY_PLAYBOOK_PROFILE.get(token, "")


def coa_types_for_playbook_profile(playbook_profile: str | None) -> tuple[str, ...] | None:
    token = (playbook_profile or "").strip().lower()
    return COA_TYPES_BY_PLAYBOOK_PROFILE.get(token)


def resolve_coa_account_name(
    suggested: str,
    accounts: list[ChartOfAccountEntry],
) -> str:
    """Match a suggested ledger name against tenant chart of accounts."""
    needle = (suggested or "").strip()
    if not needle or not accounts:
        return ""
    for entry in accounts:
        if entry.name == needle:
            return entry.name
    lowered = needle.lower()
    for entry in accounts:
        if entry.name.lower() == lowered:
            return entry.name
    for entry in accounts:
        name_lower = entry.name.lower()
        if lowered in name_lower or name_lower in lowered:
            return entry.name
    return ""


def _resolve_sparse_revenue_ledger(accounts: list[ChartOfAccountEntry]) -> str:
    revenue = [entry for entry in accounts if (entry.type or "") == "Revenue"]
    if len(revenue) == 1:
        return revenue[0].name
    import re

    hint = re.compile(r"\b(sales|revenue|income|turnover)\b", re.I)
    for entry in revenue:
        if hint.search(entry.name or ""):
            return entry.name
    return ""


def default_post_to_ledger(
    playbook_profile: str | None,
    accounts: list[ChartOfAccountEntry],
    *,
    route_target: str | None = None,
) -> str:
    suggested = suggested_ledger_for_playbook_profile(playbook_profile)
    if suggested:
        resolved = resolve_coa_account_name(suggested, accounts)
        if resolved:
            return resolved
    profile = (playbook_profile or "").strip().lower()
    route = (route_target or "").strip()
    if profile == "ar_goods" or route == "Sales Management":
        return _resolve_sparse_revenue_ledger(accounts)
    return ""

File: frontend/src/lib/documentTypeGlDefaults.ts
import type { PlaybookProfile } from "@/lib/documentPlaybookConfig";
import type { ChartOfAccountRow } from "@/api/types";
import { filterCoaAccountsForPostingRole, resolveCoaAccountName } from "@/lib/coaAccountOptions";

export const SUGGESTED_LEDGER_BY_PLAYBOOK_PROFILE: Partial<Record<PlaybookProfile | string, string>> = {
  po_goods: "Raw Materials",
  import_dossier: "Raw Materials",
  freight_logistics: "Logistics",
  standard_transactional: "Operating Expenses",
  credit_adjustment: "Operating Expenses",
  debit_note: "Operating Expenses",
  direct_expense: "Operating Expenses",
  employee_claim: "Travel Expense",
  ar_goods: "Operating Revenue",
  pre_transactional: "Suspense Account",
};

const REVENUE_NAME_HINT = /\b(sales|revenue|income|turnover)\b/i;

function resolveSparseRevenueLedger(accounts: ChartOfAccountRow[]): string {
  const revenueAccounts = filterCoaAccountsForPostingRole(accounts, "revenue");
  if (revenueAccounts.length === 1) return revenueAccounts[0].name;
  const preferred = revenueAccounts.find((row) => REVENUE_NAME_HINT.test(row.name));
  return preferred?.name ?? "";
}

export function suggestedLedgerForPlaybookProfile(playbookProfile: string | undefined): string {
  const token = (playbookProfile ?? "").trim().toLowerCase();
  return SUGGESTED_LEDGER_BY_PLAYBOOK_PROFILE[token] ?? "";
}

export function defaultPostToLedger(
  playbookProfile: string | undefined,
  accounts: ChartOfAccountRow[],
  routeTarget?: string
): string {
  const profile = (playbookProfile ?? "").trim().toLowerCase();
  const route = (routeTarget ?? "").trim();
  const suggested = suggestedLedgerForPlaybookProfile(playbookProfile);
  if (suggested) {
    const resolved = resolveCoaAccountName(suggested, accounts);
    if (resolved) return resolved;
  }
  if (profile === "ar_goods" || route === "Sales Management") {
    return resolveSparseRevenueLedger(accounts);
  }
  return "";
}

export function applyDefaultPostToIfEmpty(
  postTo: { ledger: string; subLedger: string },
  playbookProfile: string | undefined,
  accounts: ChartOfAccountRow[],
  routeTarget?: string
): { ledger: string; subLedger: string } {
  if (postTo.ledger.trim()) return postTo;
  const ledger = defaultPostToLedger(playbookProfile, accounts, routeTarget);
  return ledger ? { ...postTo, ledger } : postTo;
}

export function defaultSalesRulePostTo(accounts: ChartOfAccountRow[]): {
  ledger: string;
  subLedger: string;
  taxAccount: string;
  receivableAccount: string;
} {
  const taxAccounts = filterCoaAccountsForPostingRole(accounts, "tax_collected");
  const receivableAccounts = filterCoaAccountsForPostingRole(accounts, "receivable");
  return {
    ledger: defaultPostToLedger("ar_goods", accounts),
    subLedger: "",
    taxAccount: taxAccounts[0]?.name ?? "",
    receivableAccount: receivableAccounts[0]?.name ?? "",
  };
}
